File: test/c++/gfs/gf_chebyshev.cpp
// Copyright (c) 2025 Simons Foundation
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You may obtain a copy of the License at
//     https://www.gnu.org/licenses/gpl-3.0.txt
//
// Authors: Nils Wentzell

#include <triqs/test_tools/gfs.hpp>

// Test function: smooth function for interpolation
double test_function(double tau, double beta, double omega) {
  // Fermi function shifted to interval [0, beta]
  return 1.0 / (1.0 + std::exp(omega * (tau - beta / 2.0)));
}

TEST(GfChebyshev, Mesh) {
  double const beta = 10.0;
  auto const N      = 20;
  auto mesh_c       = mesh::chebyshev(beta, Fermion, N);

  std::cout << mesh_c << "\n";
  std::cout << "Size " << mesh_c.size() << "\n";

  for (const auto &mp : mesh_c) { std::cout << "index: " << mp.index() << ", tau: " << mp.value() << "\n"; }

  EXPECT_EQ(mesh_c.size(), N);
  EXPECT_EQ(mesh_c.beta(), beta);
  EXPECT_EQ(mesh_c.statistic(), Fermion);
}

TEST(GfChebyshev, ScalarValued) {
  double const beta  = 10.0;
  double const omega = 1.0;
  auto const N       = 30;
  auto mesh_c        = mesh::chebyshev(beta, Fermion, N);

  auto g = gf<chebyshev, scalar_valued>{mesh_c};

  // Fill with test function values at Chebyshev points
  for (auto const &mp : mesh_c) { g[mp.data_index()] = test_function(mp.value(), beta, omega); }

  // Test interpolation at various tau values
  auto tau_test  = mesh::imtime(beta, Fermion, 50);
  double max_err = 0.0;
  for (auto t : tau_test) {
    double exact  = test_function(t.value(), beta, omega);
    auto interp   = g(t);
    double err    = std::abs(exact - std::real(interp));
    max_err       = std::max(max_err, err);
  }

  std::cout << "Max interpolation error: " << max_err << "\n";
  EXPECT_LT(max_err, 1e-7); // Chebyshev should have exponential convergence
}

TEST(GfChebyshev, MatrixValued) {
  double const beta = 5.0;
  auto const N      = 15;

  auto mesh_c = mesh::chebyshev(beta, Fermion, N);
  auto g      = gf<chebyshev, matrix_valued>{mesh_c, {2, 2}};
  g()         = 0.0;

  // Set a simple linear function: g(tau) = tau * I
  for (auto const &mp : mesh_c) { g[mp.data_index()] = nda::eye<double>(2) * mp.value(); }

  // Check interpolation at midpoint
  double tau_mid = beta / 2.0;
  auto g_mid     = g(tau_mid);
  EXPECT_NEAR(std::real(g_mid(0, 0)), tau_mid, 1e-10);
  EXPECT_NEAR(std::real(g_mid(1, 1)), tau_mid, 1e-10);
  EXPECT_NEAR(std::real(g_mid(0, 1)), 0.0, 1e-14);
  EXPECT_NEAR(std::real(g_mid(1, 0)), 0.0, 1e-14);
}

TEST(GfChebyshev, InterpolationAtMeshPoints) {
  double const beta = 10.0;
  auto const N      = 20;
  auto mesh_c       = mesh::chebyshev(beta, Fermion, N);

  auto g = gf<chebyshev, scalar_valued>{mesh_c};

  // Fill with arbitrary values
  for (auto const &mp : mesh_c) { g[mp.data_index()] = mp.value() * mp.value(); }

  // Interpolation at mesh points should be exact (to machine precision)
  for (auto const &mp : mesh_c) {
    double exact = mp.value() * mp.value();
    auto interp  = g(mp.value());
    EXPECT_NEAR(exact, std::real(interp), 1e-12);
  }
}

TEST(GfChebyshev, InterpolationConvergence) {
  double const beta  = 10.0;
  double const omega = 2.0;

  std::vector<double> errors;

  for (int N : {10, 15, 20, 25, 30}) {
    auto mesh_c = mesh::chebyshev(beta, Fermion, N);
    auto g      = gf<chebyshev, scalar_valued>{mesh_c};

    for (auto const &mp : mesh_c) { g[mp.data_index()] = test_function(mp.value(), beta, omega); }

    // Test at a few points
    double max_err = 0.0;
    for (int i = 1; i < 20; ++i) {
      double tau   = beta * i / 20.0;
      double exact = test_function(tau, beta, omega);
      auto interp  = g(tau);
      max_err      = std::max(max_err, std::abs(exact - std::real(interp)));
    }
    errors.push_back(max_err);
    std::cout << "N = " << N << ", max error = " << max_err << "\n";
  }

  // Error should decrease (exponential convergence)
  for (size_t i = 1; i < errors.size(); ++i) { EXPECT_LT(errors[i], errors[i - 1]); }
}

TEST(GfChebyshev, HDF5) {
  double const beta = 10.0;
  auto const N      = 20;

  auto mesh_c = mesh::chebyshev(beta, Fermion, N);
  auto g1     = gf<chebyshev, scalar_valued>{mesh_c};
  for (auto const &mp : mesh_c) { g1[mp.data_index()] = mp.value() * mp.value(); }

  // Write and read
  auto g2 = rw_h5(g1, "test_gf_chebyshev", "g");

  EXPECT_EQ(g1.mesh(), g2.mesh());
  EXPECT_ARRAY_NEAR(g1.data(), g2.data());
}

MAKE_MAIN;

File: test/python/base/gf_chebyshev.py
# Copyright (c) 2025 Simons Foundation
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You may obtain a copy of the License at
#     https://www.gnu.org/licenses/gpl-3.0.txt
#
# Authors: Nils Wentzell

import unittest
import numpy as np
from h5 import HDFArchive
import os

from triqs.gfs import Gf
from triqs.mesh import MeshChebyshev


def test_function(tau, beta, omega):
    """Smooth test function for interpolation"""
    return 1.0 / (1.0 + np.exp(omega * (tau - beta / 2.0)))


class test_chebyshev_mesh(unittest.TestCase):

    def test_mesh_basics(self):
        """Test basic mesh construction and properties"""
        beta = 10.0
        N = 20
        mesh = MeshChebyshev(beta, 'Fermion', N)

        self.assertEqual(len(mesh), N)
        self.assertEqual(mesh.beta, beta)
        self.assertEqual(mesh.statistic, 'Fermion')

        # Check points are in [0, beta]
        for mp in mesh:
            self.assertGreaterEqual(mp.value, 0.0)
            self.assertLessEqual(mp.value, beta)

        # Check ascending order
        values = [mp.value for mp in mesh]
        self.assertEqual(values, sorted(values))

    def test_mesh_points_property(self):
        """Test access to Chebyshev points and weights"""
        beta = 10.0
        N = 15
        mesh = MeshChebyshev(beta, 'Fermion', N)

        points = mesh.points
        weights = mesh.weights

        self.assertEqual(len(points), N)
        self.assertEqual(len(weights), N)

        # Points should be in [0, beta]
        self.assertTrue(np.all(points >= 0))
        self.assertTrue(np.all(points <= beta))

    def test_gf_scalar_valued(self):
        """Test scalar-valued GF on Chebyshev mesh"""
        beta = 10.0
        N = 25
        mesh = MeshChebyshev(beta, 'Fermion', N)

        g = Gf(mesh=mesh, target_shape=[])

        # Fill with simple linear function f(tau) = tau
        for mp in mesh:
            g[mp] = mp.value

        # Test interpolation at midpoint
        tau_mid = beta / 2.0
        self.assertAlmostEqual(np.real(g(tau_mid)), tau_mid, places=10)

    def test_gf_matrix_valued(self):
        """Test matrix-valued GF on Chebyshev mesh"""
        beta = 5.0
        N = 15
        mesh = MeshChebyshev(beta, 'Fermion', N)

        g = Gf(mesh=mesh, target_shape=[2, 2])
        g.data[:] = 0.0

        # Fill with g(tau) = tau * I
        for mp in mesh:
            g[mp] = np.eye(2) * mp.value

        # Test interpolation
        tau_mid = beta / 2.0
        g_mid = g(tau_mid)
        np.testing.assert_almost_equal(np.real(g_mid[0, 0]), tau_mid, decimal=10)
        np.testing.assert_almost_equal(np.real(g_mid[1, 1]), tau_mid, decimal=10)

    def test_interpolation_accuracy(self):
        """Test interpolation accuracy with smooth function"""
        beta = 10.0
        omega = 1.0
        N = 30
        mesh = MeshChebyshev(beta, 'Fermion', N)

        g = Gf(mesh=mesh, target_shape=[])

        # Fill with test function values
        for mp in mesh:
            g[mp] = test_function(mp.value, beta, omega)

        # Test interpolation at various tau values
        tau_test = np.linspace(0.01, beta - 0.01, 50)
        max_err = 0.0
        for tau in tau_test:
            exact = test_function(tau, beta, omega)
            interp = np.real(g(tau))
            err = abs(exact - interp)
            max_err = max(max_err, err)

        # Chebyshev should have good convergence
        self.assertLess(max_err, 1e-7)

    def test_interpolation_at_mesh_points(self):
        """Test that interpolation at mesh points is exact"""
        beta = 10.0
        N = 20
        mesh = MeshChebyshev(beta, 'Fermion', N)

        g = Gf(mesh=mesh, target_shape=[])

        # Fill with arbitrary values
        for mp in mesh:
            g[mp] = mp.value ** 2

        # Interpolation at mesh points should be exact
        for mp in mesh:
            exact = mp.value ** 2
            interp = np.real(g(mp.value))
            self.assertAlmostEqual(exact, interp, places=12)

    def test_interpolation_convergence(self):
        """Test exponential convergence with increasing N"""
        beta = 10.0
        omega = 2.0

        errors = []
        for N in [10, 15, 20, 25, 30]:
            mesh = MeshChebyshev(beta, 'Fermion', N)
            g = Gf(mesh=mesh, target_shape=[])

            for mp in mesh:
                g[mp] = test_function(mp.value, beta, omega)

            # Test at a few points
            max_err = 0.0
            for i in range(1, 20):
                tau = beta * i / 20.0
                exact = test_function(tau, beta, omega)
                interp = np.real(g(tau))
                max_err = max(max_err, abs(exact - interp))
            errors.append(max_err)

        # Error should decrease
        for i in range(1, len(errors)):
            self.assertLess(errors[i], errors[i - 1])

    def test_hdf5(self):
        """Test HDF5 serialization"""
        beta = 10.0
        N = 20
        mesh = MeshChebyshev(beta, 'Fermion', N)
        g = Gf(mesh=mesh, target_shape=[])

        for mp in mesh:
            g[mp] = mp.value ** 2

        fname = 'test_chebyshev_py.h5'
        with HDFArchive(fname, 'w') as ar:
            ar['g'] = g

        with HDFArchive(fname, 'r') as ar:
            g2 = ar['g']

        self.assertEqual(g.mesh, g2.mesh)
        np.testing.assert_allclose(g.data, g2.data)

        os.remove(fname)

    def test_mesh_equality(self):
        """Test mesh equality comparison"""
        m1 = MeshChebyshev(10, 'Fermion', 20)
        m2 = MeshChebyshev(10, 'Fermion', 20)
        m3 = MeshChebyshev(10, 'Fermion', 15)
        m4 = MeshChebyshev(5, 'Fermion', 20)
        m5 = MeshChebyshev(10, 'Boson', 20)

        self.assertEqual(m1, m2)
        self.assertNotEqual(m1, m3)  # Different N
        self.assertNotEqual(m1, m4)  # Different beta
        self.assertNotEqual(m1, m5)  # Different statistics


if __name__ == '__main__':
    unittest.main()
